File: apps/api-rs/src/rag/embedding.rs
use std::time::Duration;

use anyhow::{anyhow, bail, Result};
use serde::{Deserialize, Serialize};

use crate::config::EmbeddingConfig;

#[derive(Debug, Clone)]
pub struct EmbeddingClientConfig {
    pub base_url: String,
    pub api_key: String,
    pub model: String,
    pub batch_size: usize,
    pub timeout_seconds: u64,
    pub retry_max: usize,
}

impl TryFrom<&EmbeddingConfig> for EmbeddingClientConfig {
    type Error = anyhow::Error;

    fn try_from(config: &EmbeddingConfig) -> Result<Self> {
        let api_key = config.api_key.clone().ok_or_else(|| {
            anyhow!("embedding api key is missing: set EMBED_API_KEY, LLM_API, or LLM_API_KEY")
        })?;
        if config.model.trim().is_empty() {
            bail!("embedding model is empty");
        }
        if config.base_url.trim().is_empty() {
            bail!("embedding base url is empty");
        }
        Ok(Self {
            base_url: config.base_url.clone(),
            api_key,
            model: config.model.clone(),
            batch_size: config.batch_size.clamp(1, 100),
            timeout_seconds: 120,
            retry_max: config.retry_max.max(1) as usize,
        })
    }
}

#[derive(Debug, Clone)]
pub struct EmbeddingClient {
    config: EmbeddingClientConfig,
    http: reqwest::Client,
}

#[derive(Debug, Serialize)]
struct EmbeddingRequest<'a> {
    model: &'a str,
    input: &'a [String],
}

#[derive(Debug, Deserialize)]
struct EmbeddingResponse {
    data: Vec<EmbeddingData>,
}

#[derive(Debug, Deserialize)]
struct EmbeddingData {
    embedding: Vec<f64>,
    index: Option<usize>,
}

impl EmbeddingClient {
    pub fn new(config: EmbeddingClientConfig) -> Result<Self> {
        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(config.timeout_seconds))
            .build()?;
        Ok(Self { config, http })
    }

    pub fn model(&self) -> &str {
        &self.config.model
    }

    pub fn batch_size(&self) -> usize {
        self.config.batch_size
    }

    pub async fn embed_one(&self, text: &str) -> Result<Vec<f64>> {
        let mut vectors = self.embed_batch(&[text.to_string()]).await?;
        vectors
            .pop()
            .ok_or_else(|| anyhow!("embedding provider returned no vector"))
    }

    pub async fn embed_batch(&self, texts: &[String]) -> Result<Vec<Vec<f64>>> {
        if texts.is_empty() {
            return Ok(Vec::new());
        }
        if texts.iter().any(|text| text.trim().is_empty()) {
            bail!("embedding input contains empty text");
        }

        let mut payload = None;
        for attempt in 1..=self.config.retry_max {
            let response = self
                .http
                .post(self.embeddings_url())
                .header("Authorization", format!("Bearer {}", self.config.api_key))
                .header("Content-Type", "application/json")
                .json(&EmbeddingRequest {
                    model: &self.config.model,
                    input: texts,
                })
                .send()
                .await;
            match response {
                Ok(response) if response.status().is_success() => {
                    payload = Some(response.json::<EmbeddingResponse>().await?);
                    break;
                }
                Ok(response) => {
                    let status = response.status();
                    let retryable = status.as_u16() == 429 || status.is_server_error();
                    let body = response
                        .text()
                        .await
                        .map(|body| body.chars().take(1_000).collect::<String>())
                        .unwrap_or_else(|_| "<failed to read response body>".to_string());
                    if !retryable || attempt == self.config.retry_max {
                        bail!("embedding provider returned HTTP {status}: {body}");
                    }
                }
                Err(error) => {
                    if attempt == self.config.retry_max {
                        return Err(error.into());
                    }
                }
            }
            let delay =
                Duration::from_millis((250_u64 * 2_u64.pow(attempt.min(6) as u32)).min(8_000));
            tokio::time::sleep(delay).await;
        }

        let payload =
            payload.ok_or_else(|| anyhow!("embedding retry loop completed without a response"))?;
        if payload.data.len() != texts.len() {
            bail!(
                "embedding provider returned {} vectors for {} inputs",
                payload.data.len(),
                texts.len()
            );
        }

        let mut ordered = vec![Vec::new(); texts.len()];
        for (fallback_index, item) in payload.data.into_iter().enumerate() {
            let index = item.index.unwrap_or(fallback_index);
            if index >= ordered.len() {
                bail!("embedding provider returned out-of-range index {index}");
            }
            if item.embedding.is_empty() {
                bail!("embedding provider returned empty vector at index {index}");
            }
            ordered[index] = item.embedding;
        }
        if ordered.iter().any(|v| v.is_empty()) {
            bail!("embedding provider response missed one or more vectors");
        }
        Ok(ordered)
    }

    fn embeddings_url(&self) -> String {
        let base = self.config.base_url.trim_end_matches('/');
        if base.ends_with("/embeddings") {
            base.to_string()
        } else {
            format!("{base}/embeddings")
        }
    }
}

File: apps/api-bun/src/rag/embedding.ts
// 移植自 apps/api-rs/src/rag/embedding.rs —— OpenAI 兼容 /v1/embeddings 客户端（batch + 重试 + 校验）
import type { EmbeddingConfig } from '../config.ts';

export interface EmbeddingClientConfig {
  baseUrl: string;
  apiKey: string;
  model: string;
  batchSize: number;
  timeoutSeconds: number;
  retryMax: number;
}

export function embeddingClientConfigFrom(config: EmbeddingConfig): EmbeddingClientConfig {
  const apiKey = config.apiKey;
  if (apiKey === null || apiKey === undefined) {
    throw new Error('embedding api key is missing: set EMBED_API_KEY, LLM_API, or LLM_API_KEY');
  }
  if (config.model.trim().length === 0) throw new Error('embedding model is empty');
  if (config.baseUrl.trim().length === 0) throw new Error('embedding base url is empty');
  return {
    baseUrl: config.baseUrl,
    apiKey,
    model: config.model,
    batchSize: Math.min(100, Math.max(1, config.batchSize)),
    timeoutSeconds: 120,
    retryMax: Math.max(1, config.retryMax),
  };
}

interface EmbeddingData {
  embedding: number[];
  index?: number;
}

interface EmbeddingResponse {
  data: EmbeddingData[];
}

function sleep(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export class EmbeddingClient {
  readonly config: EmbeddingClientConfig;

  constructor(config: EmbeddingClientConfig) {
    this.config = config;
  }

  model(): string {
    return this.config.model;
  }

  batchSize(): number {
    return this.config.batchSize;
  }

  async embedOne(text: string): Promise<number[]> {
    const vectors = await this.embedBatch([text]);
    const vector = vectors.pop();
    if (vector === undefined) throw new Error('embedding provider returned no vector');
    return vector;
  }

  async embedBatch(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) return [];
    if (texts.some((text) => text.trim().length === 0)) {
      throw new Error('embedding input contains empty text');
    }

    let payload: EmbeddingResponse | null = null;
    for (let attempt = 1; attempt <= this.config.retryMax; attempt++) {
      let response: Response;
      try {
        response = await fetch(this.embeddingsUrl(), {
          method: 'POST',
          headers: {
            Authorization: `Bearer ${this.config.apiKey}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ model: this.config.model, input: texts }),
          signal: AbortSignal.timeout(this.config.timeoutSeconds * 1000),
        });
      } catch (error) {
        if (attempt === this.config.retryMax) throw error;
        await sleep(retryDelayMillis(attempt));
        continue;
      }
      if (response.ok) {
        payload = parseEmbeddingResponse(await response.json());
        break;
      }
      const status = response.status;
      const retryable = status === 429 || status >= 500;
      const bodyText = await response.text().catch(() => '<failed to read response body>');
      const body = Array.from(bodyText).slice(0, 1_000).join('');
      if (!retryable || attempt === this.config.retryMax) {
        throw new Error(
          `embedding provider returned HTTP ${status} ${response.statusText}: ${body}`,
        );
      }
      await sleep(retryDelayMillis(attempt));
    }

    if (payload === null) {
      throw new Error('embedding retry loop completed without a response');
    }
    if (payload.data.length !== texts.length) {
      throw new Error(
        `embedding provider returned ${payload.data.length} vectors for ${texts.length} inputs`,
      );
    }

    const ordered: number[][] = Array.from({ length: texts.length }, () => []);
    for (const [fallbackIndex, item] of payload.data.entries()) {
      const index = item.index ?? fallbackIndex;
      if (index >= ordered.length || index < 0) {
        throw new Error(`embedding provider returned out-of-range index ${index}`);
      }
      if (item.embedding.length === 0) {
        throw new Error(`embedding provider returned empty vector at index ${index}`);
      }
      ordered[index] = item.embedding;
    }
    if (ordered.some((vector) => vector.length === 0)) {
      throw new Error('embedding provider response missed one or more vectors');
    }
    return ordered;
  }

  embeddingsUrl(): string {
    const base = this.config.baseUrl.replace(/\/+$/, '');
    if (base.endsWith('/embeddings')) return base;
    return `${base}/embeddings`;
  }
}

function retryDelayMillis(attempt: number): number {
  return Math.min(250 * 2 ** Math.min(attempt, 6), 8_000);
}

function parseEmbeddingResponse(value: unknown): EmbeddingResponse {
  if (typeof value !== 'object' || value === null) {
    throw new Error('embedding provider returned an invalid JSON response');
  }
  const data = (value as Record<string, unknown>).data;
  if (!Array.isArray(data)) {
    throw new Error('embedding provider returned an invalid JSON response');
  }
  const items: EmbeddingData[] = [];
  for (const item of data) {
    if (typeof item !== 'object' || item === null) {
      throw new Error('embedding provider returned an invalid JSON response');
    }
    const record = item as Record<string, unknown>;
    if (!Array.isArray(record.embedding) || record.embedding.some((v) => typeof v !== 'number')) {
      throw new Error('embedding provider returned an invalid JSON response');
    }
    const index = record.index;
    if (index !== undefined && (typeof index !== 'number' || !Number.isInteger(index))) {
      throw new Error('embedding provider returned an invalid JSON response');
    }
    items.push({ embedding: record.embedding as number[], ...(index !== undefined ? { index: index as number } : {}) });
  }
  return { data: items };
}
